Profile redirect notes

Gargron opened the pull request:
See also #177 (partial)

![image](https://user-images.githubusercontent.com/184731/32976524-e81fabca-cc17-11e7-93e5-7dd82e6658d6.png)

![image](https://user-images.githubusercontent.com/184731/32983119-2090bd8c-cc90-11e7-9a75-6a20aa3489d9.png)

- [x] Database structure to store information about where account moved
- [x] Serialize this information in REST API and ActivityPub
- [x] Deserialize this information from ActivityPub
- [x] Display note on public profiles
- [x] Display note on profiles in web UI

---

Please mind that this PR does not include the automatic followers unfollow/re-follow of new account, nor does it include a way to actually set this "move" information, yet. Therefore, it's only one step towards solving #177

ThibG commented on app/lib/activitypub/adapter.rb:
```diff
@@ -9,6 +9,7 @@ class ActivityPub::Adapter < ActiveModelSerializers::Adapter::Base
       {
         'manuallyApprovesFollowers' => 'as:manuallyApprovesFollowers',
         'sensitive'                 => 'as:sensitive',
+        'movedTo'                   => 'as:movedTo',
```
Should you really use ActivityStreams' namespace for that? I couldn't find any documentation for “movedTo”

Gargron replied:
I made it up since it doesn't exist, but probably should exist. @cwebber is okay with it

ThibG replied:
Alright, that's fine with me.

ThibG commented:
I'm wondering if it would make sense to redirect webfinger queries, too.

Gargron commented:
@ThibG 

> Gargron - Yesterday at 9:56 PM
@nightpool i just implemented a webfinger redirect for accounts and getting second thoughts. maybe the only redirect should be the public profile page. because a webfinger redirect would brick any look-up of all old content... plus it's a rabbit hole, there's a ton of ActivityPub endpoints, it's unclear whether you should be able to redirect them and them still be considered valid, and there's no way to tell if the new server has the same ActivityPub URI structure, etc. and so that would also brick stuff. redirecting public profile page and adding a "redirects" to the REST API so apps can show a special note, maybe that's all
nightpool - Yesterday at 10:01 PM
hm.
I don't see how redirecting the webfinger will break lookup of old content?
Gargron - Yesterday at 10:04 PM
it's all very confusing but say I am looking up "Hello world" by Gargron@mastodon.social, but Gargron@mastodon.social redirects to gargron@gonext.gg. Let's say we modified our webfinger code to even follow redirects in the first place. Now it creates "Hello world" as created by gargron@gonext.gg. You favourite it. That activity is sent to gargron@gonext.gg... but that post does not exist there, so it blackholes
That's just one example
nightpool - Yesterday at 10:04 PM
How are you looking up that post, by URI?
Gargron - Yesterday at 10:05 PM
Webfinger code not supporting redirects in all versions is even bigger. It'll just mean the content will not be processed in the first place.
nightpool - Yesterday at 10:05 PM
It should still have mastodon.social/users/gargron as the post actor
yeah the redirect is maybe a bigger concern.....
Gargron - Yesterday at 10:05 PM
It doesn't matter how you look up the post, when it checks the author it does webfinger.
nightpool - Yesterday at 10:07 PM
I don't think that's true
FetchRemoteStatusService uses attributedTo.
Gargron - Yesterday at 10:10 PM
when mastodon looks up an activitypub actor for the first time, i.e. there's no existing db record, it performs a webfinger verification
which leads to the scenario i described
there's also the question of... what if the record exists, but then webfinger cache is stale so we webfinger again, and it redirects. wtf do then
nightpool - Yesterday at 10:11 PM
Sure, okay
Gargron - Yesterday at 10:12 PM
thinking about all the ways in which this sort of thing can fuck up things is making my brain hurt

ThibG commented:
@Gargron yeah, indeed, that's what I was thinking. Since the content is still there, it kind of makes sense to not stop it from federating. The advantage is that it would be completely transparent to whichever software, OStatus or ActivityPub, supporting “movedTo” or not. The downside is that it would stop the content of the old profile from federating anymore, although it's still there… it's probably not worth it, I guess.

ThibG commented:
Another thing I'm wondering about and you haven't brought up in this PR is make every follower of the old account automatically send a follow request to the new one (and potentially unfollow the old one) once it's discovered to have moved.

Gargron commented:
@ThibG No that's not at all included in this PR.

ThibG commented:
Yes, I know, I was simply wondering if doing that would make sense.

ThibG commented on app/javascript/mastodon/features/account_timeline/components/moved_note.js:
```diff
@@ -0,0 +1,48 @@
+      <div className='account__moved-note'>
+        <div className='account__moved-note__message'>
+          <div className='account__moved-note__icon-wrapper'><i className='fa fa-fw fa-suitcase account__moved-note__icon' /></div>
+          <FormattedMessage id='account.moved_to' defaultMessage='{name} has moved to:' values={{ name: <strong dangerouslySetInnerHTML={displayNameHtml} /> }} />
```
Wouldn't it better if the name was a link to the new account?

Gargron replied:
That's the old name

Cam-B commented:
One step further. Awesome!
